### PR TITLE
Add write-config-to to scheduler

### LINKED ISSUE #58805: kube-scheduler doesn not have --write-config-to option
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
We face
```
WARNING: all flags than --config are deprecated. Please begin using a config file ASAP.
```
But there is no way to generate default config file

**How to reproduce it (as minimally and precisely as possible)**:
```
kube-scheduler --write-config-to ./scheduler.conf
Error: unknown flag: --write-config-to
```

**Environment**:
- Kubernetes version 1.9.2

### PR DESCRIPTION
**Pull request by resouer**
**What this PR does / why we need it**:
Scheduler should be able to write its default configure to file. This actually applies to all components which claims options other than `--config` will be deprecated.

Otherwise, users will be super confused to find out how to write a proper config file to these components.

See: https://stackoverflow.com/questions/47966440/how-to-create-a-config-file-for-kube-scheduler-to-use-by-the-config-argument
ref: #52562

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #58805

Usage:

```bash
./_output/bin/kube-scheduler --write-config-to /tmp/kube-scheduler.yaml
```

**Special notes for your reviewer**:
This should have been fixed several releases ago, so lets include it in 1.11

**Release note**:

```release-note
Add write-config-to to scheduler
```

### CODE REVIEW

**Comment by resouer** [comment]
cc @corest should be interested in

**Review by xiangpengzhao — COMMENTED** [review]
I like this idea, but maybe we can just document an straightforward example somewhere rather than coding ?

File: cmd/kube-scheduler/app/server.go [hunk 1]
```diff
@@ -167,8 +172,8 @@ func NewOptions() (*Options, error) {
 }
 
 func (o *Options) Complete() error {
-	if len(o.ConfigFile) == 0 {
-		glog.Warning("WARNING: all flags other than --config are deprecated. Please begin using a config file ASAP.")
+	if len(o.ConfigFile) == 0 && len(o.WriteConfigTo) == 0 {
```
[the remark this answers is not in the document]
Comment (resouer) [reply cmd/kube-scheduler/app/server.go]: Nope. It's just a warning.

File: cmd/kube-scheduler/app/server.go [hunk 2]
```diff
@@ -106,6 +110,7 @@ type Options struct {
 // AddFlags adds flags for a specific SchedulerServer to the specified FlagSet
 func (o *Options) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&o.ConfigFile, "config", o.ConfigFile, "The path to the configuration file.")
+	fs.StringVar(&o.WriteConfigTo, "write-config-to", o.WriteConfigTo, "If set, write the default configuration values to this file and exit.")
```
[the remark this answers is not in the document]
Comment (resouer) [reply cmd/kube-scheduler/app/server.go]: Actually, this patch always generate default values, which is expected.

**Comment by resouer** [comment]
> I like this idea, but maybe we can just document an straightforward example somewhere rather than coding ?

Actually as we are migrating to `ComponentConfig` , `--write-config-to` will become a mandatory option and it has already been added to several components like kube-proxy I believe.

File: cmd/kube-scheduler/app/server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (xiangpengzhao) [reply cmd/kube-scheduler/app/server.go]: Yeah you're right. I mis-read some codes.

File: cmd/kube-scheduler/app/server.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (xiangpengzhao) [reply cmd/kube-scheduler/app/server.go]: But `o.applyDeprecatedHealthzAddressToConfig()` and friends aren't just warning. They apply `healthzAddress` (which is the flag `--address`), etc. So if users specify flags like this `--write-config-to ./scheduler.yaml --address 1.2.3.4`, the flag `--address 1.2.3.4` won't take effect. Maybe I miss something here? :)

[the remark this answers is not in the document]
Comment (xiangpengzhao) [reply cmd/kube-scheduler/app/server.go]: Mis-read again... When `--write-config-to` is specified, kube-scheduler will write the config and exit. So it's ok as-is.

File: cmd/kube-scheduler/app/server.go (hunk 2 above)
Comment (k82cn) [inline cmd/kube-scheduler/app/server.go]: If the `config` is not nil, the value is not default, right?

Comment (resouer) [reply cmd/kube-scheduler/app/server.go]: No, only write default values from schema, which is my intention.

**Comment by resouer** [comment]
@bsalamat @k82cn Currently, I intentionally make it write default values to file only, but please let me know if you think we should write out real values in `--config` if provided.

**Comment by k82cn** [comment]
> So my 2c is let's keep current design and I can make those two flags exclusive. After all, it does make sense to use them together.

That's ok to me. If writing out real values is **NOT** complex, it seems better as it cover more case (writting default value if no `--config`).

**Comment by resouer** [comment]
@k82cn There's no tech blocker. Just hope to make sure this alias with users expectation.

@AshleyByeUK @tiukhtinvladimir I saw you are tracking this issue, do you expect `--write-config-to` generate user set values when used with `--config`, or it's better to stick it to always default. Or you just don't care too much?

**Comment by kennethredler** [comment]
@resouer IMO, do default first. Then iterate to include `--config` values on next round.

**Comment by AshleyByeUK** [comment]
@resouer I think it makes sense for `--write-config-to` to provide default values. If `--config` flag is applied, it would seem more appropriate to write out the current config - this latter flag is likely to only be an interim requirement to allow existing command line arguments to be converted to a config file, right?

Ultimately, I'm not overly fussed, but this seems like the sanest approach.

**Comment by resouer** [comment]
@AshleyByeUK Correct.

Thanks guys. According to current discussion, I will make `-write-config-to` write out default values, but if `--config` is provided at the same time, it will write out given values.

**Comment by resouer** [comment]
@k82cn Take a final look?

**Pull request merged**